README
README
vPICdisasm - Version 1.3 - 04/03/2011
Vanya A. Sergeev - <vsergeev@gmail.com>
================================================================================
Table of Contents

1. ABOUT vPICdisasm
2. LICENSE
3. COMPILING vPICdisasm
4. USING vPICdisasm
5. Ghetto Address Labels
6. Shortcomings
7. Source Code
8. Sample Disassembly Outputs  
================================================================================

1. ABOUT vPICdisasm
================================================================================
vPICdisasm is a Microchip PIC firmware disassembler that supports the Baseline,
Mid-Range, and Mid-Range Enhanced 8-bit PIC cores. This single-pass disassembler
can read Intel HEX8 and Motorola S-Record formatted files containing valid PIC
program binaries.

vPICdisasm fully supports all 35 Mid-Range PIC instructions (as well as the two
deprecated ones: "option" and "tris"), the additional 21 Mid-Range Enhanced 
PIC instructions, and the 33 Baseline PIC instructions.

vPICdisasm features a handful of formatting options, including:
 - Printing the instruction addresses alongside disassembly, enabled by default
 - Printing the destination address of relative branch/jump/call instructions
   as comments alongside disassembly, enabled by default
 - Printing the original opcode data alongside disassembly
 - Ghetto Address Labels (see "Ghetto Address Labels" section)
 - Literal operands represented in either hexadecimal, binary, or decimal bases,
   and as ASCII in an assembly comment 
 - data word directive for data not recognized as an instruction during
   disassembly
 - Piped input and output

vPICdisasm should work on most *nix platforms, including a Cygwin or MinGW
environment. vPICdisasm was written by Vanya A. Sergeev, and tested with the
GNU C Compiler on Linux. Feel free to send any ideas or suggestions to vsergeev
at gmail dot com.

2. LICENSE
================================================================================
vPICdisasm is released under the GNU General Public License Version 2.
    You should have received a copy of the GNU General Public License
    along with this program; see the file "COPYING".  If not, visit
    http://www.gnu.org or write to the Free Software Foundation, Inc.,
    59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

3. COMPILING vPICdisasm
================================================================================
Simply by running,
$ make
 in the vPICdisasm project directory should compile vPICdisasm on most
*nix systems, including a Cygwin or MinGW environment. The Makefile is
configured to use GCC to compile vPICdisasm.

vPICdisasm should have no problem being compiled with "gmake".

4. USING vPICdisasm
================================================================================

* File Input:
	For most purposes:
	 $ ./vpicdisasm <PIC program file>
	Example:
	 $ ./vpicdisasm sampleprogram.hex

	Use - for standard input.

	vPICdisasm will assume the Mid-Range PIC architecture by default. You can
	specifiy an alternate architecture with the -a or --architecture option.
	Example:
	$ ./vpicdisasm -a enhanced sampleprogram.hex

* Option -t or --file-type
	vPICdisasm will auto-recognize Intel HEX8, and Motorola S-Record files
	by their first character. However, the -t or --file-type option can be
	used to explicitly select the file format.
	Example:
	 $ ./vpicdisasm -t ihex sampleprogram

	The file type argument for this option can be "ihex", or "srecord" for 
	Intel HEX8 or Motorola S-Record formatted files, respectively.

* Option -o or --out-file <output file>
	Specify an output file for writing instead of the standard output. The
	output file - is also synonymous for standard output.

* Option -a or --architecture
	vPICdisasm supports the Baseline, Mid-Range, and Mid-Range Enhanced
	8-bit PIC cores. The architecture can be specified with this option,
	followed by the architecture identifier (right side of the list below):
		Baseline		baseline
		Mid-Range		midrange (default)
		Mid-Range Enhanced	enhanced

* Option --original
	Print the original opcode data to the left of the disassembly.
	Note: this option is ignored if address labels are enabled (to ensure
	assemble-able code).

* Options --literal-hex, --literal-bin, --literal-dec
	vPICdisasm can represent literal operands in either hexadecimal, 
	decimal, or binary bases. The base can be specified with the
	--literal-hex, --literal-bin, and --literal-dec options.

* Options --literal-ascii
	Display the ASCII value of a literal operand in an assembly comment.

* Options --no-addresses, --no-destination-comments
	By default, vPICdisasm will print the instruction addresses alongside 
	disassembly and destination comments for relative branch, jump, and call
	instructions. These formatting options can be disabled with the 
	--no-addresses and --no-destination-comments options, respectively.

* Options -l or --address-label	
	See the "Ghetto Address Labels" section.

* Options -h or --help, -v or --version
	The -h or --help option will print a brief usage summary, including 
	supported program options and file types.
	The -v or --version option will print the program's version number.

If you encounter any program bugs or problems, please notify the program
author by email: Vanya A. Sergeev - vsergeev at gmail dot com.

5. Ghetto Address Labels
================================================================================
vPICdisasm supports a unique formatting feature: Ghetto Address Labels, which 
few, if not any, single-pass disassemblers implement.

With the -l or --address-label option and a supplied prefix, vPICdisasm
will print a label containing the ideally non-numerical supplied prefix 
and the address of the disassembled instruction at every instruction. Also, 
all relative branch, jump, and call instructions will be formatted to jump to 
their designated address label.

This feature enables direct re-assembly of the vPICdisasm's disassembly.
This can be especially useful for quick modification of the PIC program assembly
code without having to manually format the disassembly or adjust the relative
branch, jump, or call distances with every modification to the disassembly.

The -l or --address-label option overrides the default printing of the
addresses alongside disassembly. Destination comments can still be printed.

Example:
$ ./vpicdisasm -l "A_" sampleprogram.hex
vPICdisasm's disassembly will include address labels that will look like this:
A_0000.

For sample disassembly outputs by vPICdisasm, see the "Sample Disassembly
Outputs" section.

6. Shortcomings
================================================================================
vPICdisasm does not yet support the PIC18 architecture.

vPICdisasm does not disassemble and display alternate versions of the same
encoded instruction. This technically means that some instructions can never be
displayed in the disassembly because another instruction may precede it in
priority.

These features do not affect the accuracy of the disassembler's output, and 
may be supported in future versions of vPICdisasm.

7. Source Code
================================================================================
vPICdisasm's source code is heavily commented, because this disassembler was
also a personal learning project of the author.

Operand prefixes (such as "0x" for address operand) can be customized in the 
format.h header file. 

Field width spacing of the addresses printed alongside disassembly can be 
customized in the ui.c source file.

vPICdisasm uses libGIS, a free Atmel Generic, Intel HEX8, and Motorola S-Record 
Parser Library to parse formatted files containing PIC program binaries. libGIS
is available for free under both MIT and a Public Domain licenses at:
http://dev.frozeneskimo.com/software_projects/libgis
libGIS is compiled into vPICdisasm--it does not need to be obtained separately.

8. Sample Disassembly Outputs
================================================================================
Here are a few sample disassembly outputs illustrating the various formatting
options and disassembly settings vPICdisasm supports:

$ ./vpicdisasm sampleprogram.hex 
   0:   movlw 0x0
   1:   tris 0x06
   2:   movlw 0xFF
   3:   movwf 0x06
   4:   goto 0x004

$ ./vpicdisasm --original sampleprogram.hex
   0:	30 00		movlw 0x0
   1:	00 66		tris 0x06
   2:	30 FF		movlw 0xFF
   3:	00 86		movwf 0x06
   4:	28 04		goto 0x004

$ ./vpicdisasm --no-addresses sampleprogram.hex
movlw 0x0
tris 0x06
movlw 0xFF
movwf 0x06
goto 0x004

$ ./vpicdisasm --literal-bin sampleprogram.hex
   0:   movlw b'00000000'
   1:   tris 0x06
   2:   movlw b'11111111'
   3:   movwf 0x06
   4:   goto 0x004

$ ./vpicdisasm -l "A_" sampleprogram.hex

org 0x000
A_000 movlw 0x0
A_001 tris 0x06
A_002 movlw 0xFF
A_003 movwf 0x06
A_004 goto A_004
end

$ ./vpicdisasm --literal-ascii sampleprogram2.hex
   0:   retlw 0x48      ; 'H'
   1:   retlw 0x45      ; 'E'
   2:   retlw 0x4C      ; 'L'
   3:   retlw 0x4C      ; 'L'
   4:   retlw 0x4F      ; 'O'



ChangeLog
Release 1.3 - 04/03/2011
 - CRITICAL FIX: Fixed S-Record reading bug that was ignoring valid data
   records.
 - Added output file support by -o/--out-file <output file> option.
 - Added standard input support with the "-" file argument, meaning the
   disassembler now supports piped input.
 - Improved Intel HEX8 / Motorola S-Record auto-detection by first character
   rather than file extension.
 - Thanks to Thomas for all four of the above fixes and suggestions!
 - Added printing of original opcode data alongside disassembly with
   --original option.

Release 1.2 - 07/28/2010
 - Added support for Baseline and Mid-Range Enhanced PIC cores.
 - Updated destination register operand formatting to print W/F instead of 0/1.
 - Renamed source files to make more sense and for better organization of code.

Release 1.1 - 07/20/2009
 - Fixed handling of newlines (sometimes found at the end of program files) so
   an "invalid record" error doesn't appear when a newline is encountered.
 - CRITICAL FIX: Fixed reading and disassembly of odd byte length records in
   Intel HEX8 and Motorola S-Record files. Special thanks to Ahmed for
   discovery and patch!
 - Fixed a few small bugs/typos for cleaner compilation.

Release 1.0 - 01/06/2007
 - Initial release.


ui.c
/*
 * vPICdisasm - PIC program disassembler.
 * Written by Vanya A. Sergeev - <vsergeev@gmail.com>
 *
 * Copyright (C) 2007-2011 Vanya A. Sergeev
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA. 
 *
 * ui.c - Main user interface to PIC disassembler. Takes care of program arguments
 *  setting disassembly formatting options, and program file type recognition. 
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include "file.h"
#include "errorcodes.h"

/* Flags for some long options that don't have a short option equivilant */
static int no_addresses = 0;				/* Flag for --no-addresses */
static int literal_base = 0;				/* Value of the literal base (hex, bin, dec) */
static int literal_ascii_comment = 0;			/* Flag for --literal-ascii */
static int no_destination_comments = 0;			/* Flag for --no-destination-comments */
static int original_opcode = 0;				/* Flag for --original */

static struct option long_options[] = {
	{"address-label", required_argument, NULL, 'l'},
	{"arch", required_argument, NULL, 'a'},
	{"out-file", required_argument, NULL, 'o'},
	{"file-type", required_argument, NULL, 't'},
	{"no-addresses", no_argument, &no_addresses, 1},
	{"literal-bin", no_argument, &literal_base, FORMAT_OPTION_LITERAL_BIN},
	{"literal-dec", no_argument, &literal_base, FORMAT_OPTION_LITERAL_DEC},
	{"literal-hex", no_argument, &literal_base, FORMAT_OPTION_LITERAL_HEX},
	{"literal-ascii", no_argument, &literal_ascii_comment, 1},
	{"original", no_argument, &original_opcode, 1},
	{"no-destination-comments", no_argument, &no_destination_comments, 1},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'v'},
	{NULL, 0, NULL, 0}
};

static void printUsage(FILE *stream, const char *programName) {
	fprintf(stream, "Usage: %s <option(s)> <file>\n", programName);
	fprintf(stream, " Disassembles PIC program file <file>. Use - for standard input.\n");
	fprintf(stream, " Written by Vanya A. Sergeev - <vsergeev@gmail.com>.\n\n");
	fprintf(stream, " Additional Options:\n\
  -o, --out-file <output file>	Write to output file instead of standard output.\n\
  -a, --arch <architecture>	Specify the 8-bit PIC architecture to use\n\
				during disassembly.\n\
  -t, --file-type <type>	Specify the file type of the object file.\n\
  -l, --address-label <prefix> 	Create ghetto address labels with \n\
				the specified label prefix.\n\
  --original                    Print original opcode data alongside\n\
				disassembly.\n\
  --no-addresses		Do not display the address alongside\n\
				disassembly.\n\
  --no-destination-comments	Do not display the destination address\n\
				comments of relative branch instructions.\n\
  --literal-hex			Represent literals in hexadecimal (default)\n\
  --literal-bin			Represent literals in binary\n\
  --literal-dec			Represent literals in decimal\n\
  --literal-ascii		Show ASCII value of literal operands in a\n\
				comment\n\
  -h, --help			Display this usage/help.\n\
  -v, --version			Display the program's version.\n\n");
	fprintf(stream, "Supported 8-bit PIC Architectures:\n\
  Baseline 			baseline\n\
  Mid-Range			midrange (default)\n\
  Enhanced Mid-Range		enhanced\n\n");
	fprintf(stream, "Supported file types:\n\
  Intel HEX8 			ihex\n\
  Motorola S-Record 		srecord\n\n");
}

static void printVersion(FILE *stream) {
	fprintf(stream, "vPICdisasm version 1.3 - 04/03/2011.\n");
	fprintf(stream, "Written by Vanya Sergeev - <vsergeev@gmail.com>\n");
}
	
int main(int argc, const char *argv[]) {
	int optc;
	FILE *fileIn, *fileOut;
	char arch[9], fileType[8];
	int archSelect;
	int (*disassembleFile)(FILE *, FILE *, formattingOptions, int);
	formattingOptions fOptions;

	/* Recent flag options */
	fOptions.options = 0;
	/* Set default address field width for this version. */
	fOptions.addressFieldWidth = 3;
	/* Default output file to stdout */
	fileOut = stdout;

	arch[0] = '\0';
	fileType[0] = '\0';	
	while (1) {
		optc = getopt_long(argc, (char * const *)argv, "o:a:t:l:hv", long_options, NULL);
		if (optc == -1)
			break;
		switch (optc) {
			/* Long option */
			case 0:
				break;
			case 'l':
				fOptions.options |= FORMAT_OPTION_ADDRESS_LABEL;
				strncpy(fOptions.addressLabelPrefix, optarg, sizeof(fOptions.addressLabelPrefix));
				break;
			case 't':
				strncpy(fileType, optarg, sizeof(fileType));
				break;
			case 'a':
				strncpy(arch, optarg, sizeof(arch));
				break;
			case 'o':
				if (strcmp(optarg, "-") != 0)
					fileOut = fopen(optarg, "w");
				break;
			case 'h':
				printUsage(stderr, argv[0]);
				exit(EXIT_SUCCESS);	
			case 'v':
				printVersion(stderr);
				exit(EXIT_SUCCESS);
			default:
				printUsage(stdout, argv[0]);
				exit(EXIT_SUCCESS);
		}
	}

	if (!no_addresses)
		fOptions.options |= FORMAT_OPTION_ADDRESS;
	if (!no_destination_comments)
		fOptions.options |= FORMAT_OPTION_DESTINATION_ADDRESS_COMMENT;	

	if (literal_base == FORMAT_OPTION_LITERAL_DEC)
		fOptions.options |= FORMAT_OPTION_LITERAL_DEC;
	else if (literal_base == FORMAT_OPTION_LITERAL_BIN)
		fOptions.options |= FORMAT_OPTION_LITERAL_BIN;
	else
		fOptions.options |= FORMAT_OPTION_LITERAL_HEX;

	if (literal_ascii_comment)
		fOptions.options |= FORMAT_OPTION_LITERAL_ASCII_COMMENT;

	if (original_opcode)
		fOptions.options |= FORMAT_OPTION_ORIGINAL_OPCODE;
	
	if (fileOut == NULL) {
		perror("Error: Cannot open output file for writing");
		exit(EXIT_FAILURE);
	}

	/* If there are no more arguments left */
	if (optind == argc) {
		fprintf(stderr, "Error: No program file specified! Use - for standard input.\n\n");
		printUsage(stderr, argv[0]);
		if (fileOut != stdout)
			fclose(fileOut);
		exit(EXIT_FAILURE);
	}

	/* Support reading from stdin with filename "-" */
	if (strcmp(argv[optind], "-") == 0) {
		fileIn = stdin;
	} else {
		fileIn = fopen(argv[optind], "r");
		if (fileIn == NULL) {
			perror("Error: Cannot open program file for disassembly");
			if (fileOut != stdout)
				fclose(fileOut);
			exit(EXIT_FAILURE);
		}
	}

	/* If no file type was specified, try to auto-recognize the first character of the file */
	if (fileType[0] == '\0') {
		int c;
		c = fgetc(fileIn);
		/* Intel HEX record statements start with : */
		if ((char)c == ':')
			strcpy(fileType, "ihex");
		/* Motorola S-Record record statements start with S */
		else if ((char)c == 'S')
			strcpy(fileType, "srecord");
		else {
			fprintf(stderr, "Unable to auto-recognize file type by first character.\n");
			fprintf(stderr, "Please specify file type with -t,--file-type option.\n");
			if (fileOut != stdout)
				fclose(fileOut);
			if (fileIn != stdin)
				fclose(fileIn);
			exit(EXIT_FAILURE);
		}
		ungetc(c, fileIn);
	}

	/* If no architecture was specified, use midrange by default */
	if (arch[0] == '\0') {
		archSelect = PIC_MIDRANGE;
	} else {
		if (strcasecmp(arch, "baseline") == 0)
			archSelect = PIC_BASELINE;
		else if (strcasecmp(arch, "midrange") == 0)
			archSelect = PIC_MIDRANGE;
		else if (strcasecmp(arch, "enhanced") == 0)
			archSelect = PIC_MIDRANGE_ENHANCED;
		else {
			fprintf(stderr, "Unknown 8-bit PIC architecture %s.\n", arch);
			fprintf(stderr, "See program help/usage for supported PIC architectures.\n");
			if (fileOut != stdout)
				fclose(fileOut);
			if (fileIn != stdin)
				fclose(fileIn);
			exit(EXIT_FAILURE);
		}
	}
	
	if (strcasecmp(fileType, "ihex") == 0)
		disassembleFile = disassembleIHexFile;
	else if (strcasecmp(fileType, "srecord") == 0)
		disassembleFile = disassembleSRecordFile;
	else {
		if (fileType[0] != '\0')
			fprintf(stderr, "Unknown file type %s.\n", fileType);
		else
			fprintf(stderr, "Unspecified file type.\n");
		fprintf(stderr, "See program help/usage for supported file types.\n");
		if (fileOut != stdout)
			fclose(fileOut);
		if (fileIn != stdin)
			fclose(fileIn);
		exit(EXIT_FAILURE);
	}

	disassembleFile(fileOut, fileIn, fOptions, archSelect);

	if (fileOut != stdout)	
		fclose(fileOut);
	if (fileOut != stdin)
		fclose(fileIn);

	exit(EXIT_SUCCESS);
}

